### PR TITLE
Mail upgrade

### PR DESCRIPTION
**Pull request by hschu12**
Frontend and Backend upgrade of our mail service.
PR contains all issues of mail upgrade epic  as well as a minor change that disallowes grants to move away from end states (#2147)

@Tiggles: I am pretty sure that the frontend could be better but it works.

Screenshot 2021-03-10 at 10 25 15
Screenshot 2021-03-10 at 13 52 03

### CODE REVIEW

File: backend/mail-service/src/jvmMain/kotlin/mail/services/SettingsService.kt [hunk 1]
```diff
@@ -0,0 +1,114 @@
+package mail.services
```
Comment (DanThrane) [inline backend/mail-service/src/jvmMain/kotlin/mail/services/SettingsService.kt:1]:
```suggestion
package dk.sdu.cloud.mail.services
```

File: backend/mail-service/src/jvmMain/kotlin/mail/utils/MailTemplates.kt [hunk 2]
```diff
@@ -0,0 +1,239 @@
+package mail.utils
```
Comment (DanThrane) [inline backend/mail-service/src/jvmMain/kotlin/mail/utils/MailTemplates.kt:1]:
```suggestion
package dk.sdu.cloud.mail.utils
```

File: backend/mail-service/src/jvmMain/kotlin/mail/utils/MailTemplates.kt [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+        <ul>
+            <li>PIs and admins are asked to occasionally review members of their project</li>
+            <li>We ask you to ensure that only the people who need access have access</li>
+            <li>
+                If you find someone who should not have access then remove them by clicking 'Remove'
+                next to their name
+            </li>
+            <li>
+                You can begin the review by clicking 
+                <a href="https://cloud.sdu.dk/app/projects}">here</a>.
```
Comment (DanThrane) [inline backend/mail-service/src/jvmMain/kotlin/mail/utils/MailTemplates.kt:37]:
```suggestion
                <a href="https://cloud.sdu.dk/app/projects">here</a>.
```

File: backend/mail-service/src/jvmMain/kotlin/mail/utils/MailTemplates.kt [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+        </ul>
+        $NO_NOTIFICATIONS_DISCLAIMER
+    """.trimMargin()
+
+fun resetPasswordTemplate(receiver: String, token: String) =
+    """
+        <p>Dear ${escapeHtml(receiver)}</p> 
+        <p>We have received a request to reset your UCloud account password. To proceed, follow the link below.</p>
+        <p>
+            <a href="https://cloud.sdu.dk/app/login?password-reset=true&token=${token}">https://cloud.sdu.dk/app/login?password-reset=true&token=${token}</a>
```
Comment (DanThrane) [inline backend/mail-service/src/jvmMain/kotlin/mail/utils/MailTemplates.kt:48]: Just to be on the safe side I think we should escapeHtml from the tokens.

File: backend/mail-service/src/jvmMain/resources/db/migration/V2__emailSettings.sql [hunk 5]
```diff
@@ -0,0 +1,4 @@
+CREATE TABLE email_settings(
+    username varchar(255) UNIQUE NOT NULL,
```
Comment (DanThrane) [inline backend/mail-service/src/jvmMain/resources/db/migration/V2__emailSettings.sql:2]: We don't have any hard guarantees that usernames don't exceed 255 characters.

```suggestion
    username TEXT PRIMARY KEY,
```

File: frontend-web/webclient/app/UserSettings/api.tsx [hunk 6]
```diff
@@ -20,6 +21,60 @@ export function listUserSessions(
...
+    return {
+        context: "",
+        method: "POST",
+        path: "/api/mail" + "/toggleEmailSettings",
+        parameters: request,
+        reloadId: Math.random(),
+        payload: request,
+    };
+}
+
```
Comment (DanThrane) [inline frontend-web/webclient/app/UserSettings/api.tsx:77]: Should use the code gen script instead of entering by hand.

**Review by DanThrane — CHANGES_REQUESTED** [review]
A few minor things should be changed.

It also looks like you accidentially added some new files which shouldn't be in there:

- `frontend-web/webclient/npm`
- `frontend-web/webclient/ucloud@0.42.3`
- `frontend-web/webclient/webpack`

File: frontend-web/webclient/app/UserSettings/ChangeEmailSettings.tsx [hunk 7]
```diff
@@ -0,0 +1,336 @@
...
+    newGrantApplication: true,
+    grantAutoApprove: true,
+    grantApplicationUpdated: true,
+    grantApplicationApproved: true,
+    grantApplicationRejected: true,
+    grantApplicationWithdrawn: true,
+    newCommentOnApplication: true,
+    applicationTransfer: true,
+    applicationStatusChange: true,
+    //Project
```
Comment (Tiggles) [inline frontend-web/webclient/app/UserSettings/ChangeEmailSettings.tsx:27]: Nit picking, but please add a space after `//`

File: frontend-web/webclient/app/UserSettings/ChangeEmailSettings.tsx [hunk 8]
```diff
@@ -0,0 +1,336 @@
...
+    NEW_GRANT_APPLICATION,
+    GRANT_AUTO_APPROVE,
+    GRANT_APPLICATION_UPDATED,
+    GRANT_APPLICATION_APPROVED,
+    GRANT_APPLICATION_REJECTED,
+    GRANT_APPLICATION_WITHDRAWN,
+    NEW_COMMENT_ON_APPLICATION,
+    APPLICATION_TRANSFER,
+    APPLICATION_STATUS_CHANGE,
+    //Project
```
Comment (Tiggles) [inline frontend-web/webclient/app/UserSettings/ChangeEmailSettings.tsx:46]: Same as above

File: frontend-web/webclient/app/UserSettings/ChangeEmailSettings.tsx [hunk 9]
```diff
@@ -0,0 +1,336 @@
...
+    USER_ROLE_CHANGE,
+    USER_LEFT,
+    LOW_FUNDS
+}
+
+const initialState: UserDetailsState = {
+    settings: defaultEmailSettings
+};
+
+type Action<T, B> = {type: T; payload: B};
```
Comment (Tiggles) [inline frontend-web/webclient/app/UserSettings/ChangeEmailSettings.tsx:59]: I'm pretty sure PayloadAction exists elsewhere that has the same signature.

File: frontend-web/webclient/app/UserSettings/ChangeEmailSettings.tsx [hunk 10]
```diff
@@ -0,0 +1,336 @@
...
+    }, []);
+
+    const update = useCallback(async () => {
+        dispatch({
+            type: "UpdatePlaceholdersEmailSettings",
+            payload: {
+                settings: currentEmailSettings
+            }
+        })
+    }, [])
```
Comment (Tiggles) [inline frontend-web/webclient/app/UserSettings/ChangeEmailSettings.tsx:101]: Missing `;`

File: frontend-web/webclient/app/UserSettings/ChangeEmailSettings.tsx [hunk 11]
```diff
@@ -0,0 +1,336 @@
...
+        case "UpdatePlaceholdersEmailSettings":
+            return {...state, ...action.payload};
+    }
+};
+
+export const ChangeEmailSettings: React.FunctionComponent<{setLoading: (loading: boolean) => void}> = () => {
+    const [commandLoading, invokeCommand] = useCloudCommand();
+    const [state, dispatch] = React.useReducer(reducer, initialState, () => initialState);
+
+    var currentEmailSettings: EmailSettings = defaultEmailSettings
```
Comment (Tiggles) [inline frontend-web/webclient/app/UserSettings/ChangeEmailSettings.tsx:73]: You are assigning `currentEmailSettings` to equal `defaultEmailSettings` at every component update. Is this intended?

File: frontend-web/webclient/app/UserSettings/api.tsx [hunk 12]
```diff
@@ -20,6 +21,60 @@ export function listUserSessions(
     };
 }
 
+export interface EmailSettings {
+    //Grant applications
```
Comment (Tiggles) [inline frontend-web/webclient/app/UserSettings/api.tsx:25]: Same as other complaint

File: frontend-web/webclient/app/UserSettings/api.tsx [hunk 13]
```diff
@@ -20,6 +21,60 @@ export function listUserSessions(
...
+    newGrantApplication: boolean;
+    grantAutoApprove: boolean;
+    grantApplicationUpdated: boolean;
+    grantApplicationApproved: boolean;
+    grantApplicationRejected: boolean;
+    grantApplicationWithdrawn: boolean;
+    newCommentOnApplication: boolean;
+    applicationTransfer: boolean;
+    applicationStatusChange: boolean;
+    //Project
```
Comment (Tiggles) [inline frontend-web/webclient/app/UserSettings/api.tsx:35]: Same as above

File: frontend-web/webclient/app/ui-components/ClickableDropdown.tsx [hunk 14]
```diff
@@ -7,6 +7,7 @@ import {Dropdown, DropdownContent} from "./Dropdown";
 import {PropsWithChildren, useCallback, useEffect, useMemo, useRef, useState} from "react";
 
 export interface ClickableDropdownProps<T> {
+    id?: string;
```
Comment (Tiggles) [inline frontend-web/webclient/app/ui-components/ClickableDropdown.tsx:10]: Why is this added? Console complaint?

File: frontend-web/webclient/app/UserSettings/ChangeEmailSettings.tsx [hunk 15]
```diff
@@ -0,0 +1,336 @@
...
+                currentEmailSettings.userLeft = state.settings.userLeft
+                break;
+            case MailType.LOW_FUNDS:
+                state.settings.lowFunds = !state.settings.lowFunds
+                currentEmailSettings.lowFunds = state.settings.lowFunds
+                break;
+            case MailType.APPLICATION_STATUS_CHANGE:
+                state.settings.applicationStatusChange = !state.settings.applicationStatusChange
+                currentEmailSettings.applicationStatusChange = state.settings.applicationStatusChange
+                break;
```
Comment (Tiggles) [inline frontend-web/webclient/app/UserSettings/ChangeEmailSettings.tsx:176]: You're supposed to call the reducer with the new value. I'm not sure how this works in its current state.

**Review by Tiggles — CHANGES_REQUESTED** [review]
Some changes or explanations needed before I can approve it.

File: frontend-web/webclient/app/UserSettings/ChangeEmailSettings.tsx (hunk 15 above)
Comment (DanThrane) [reply frontend-web/webclient/app/UserSettings/ChangeEmailSettings.tsx:176]: Is a reducer even needed for this use case? It seems like something which is simple enough that it could just be implemented with a `useState` hook. I am not sure I see the clear benefits of using a custom reducer here.

Comment (Tiggles) [reply frontend-web/webclient/app/UserSettings/ChangeEmailSettings.tsx:176]: If an object is used instead of a primitive value, I think the reducer is the best approach.

File: frontend-web/webclient/app/UserSettings/ChangeEmailSettings.tsx [hunk 16]
```diff
@@ -91,92 +91,87 @@ export const ChangeEmailSettings: React.FunctionComponent<{setLoading: (loading:
...
-    const update = useCallback(async () => {
+    /*const update = useCallback(async () => {
         dispatch({
             type: "UpdatePlaceholdersEmailSettings",
             payload: {
                 settings: currentEmailSettings
             }
         })
-    }, [])
+    }, []);*/
```
Comment (Tiggles) [inline frontend-web/webclient/app/UserSettings/ChangeEmailSettings.tsx]: This could be removed

File: frontend-web/webclient/app/UserSettings/ChangeEmailSettings.tsx [hunk 17]
```diff
@@ -91,92 +91,87 @@ export const ChangeEmailSettings: React.FunctionComponent<{setLoading: (loading:
...
-                currentEmailSettings.applicationStatusChange = state.settings.applicationStatusChange
+                break;
+            case MailType.APPLICATION_TRANSFER:
+                state.settings.applicationTransfer = !state.settings.applicationTransfer
                 break;
         }
+        dispatch({
+            type: "UpdatePlaceholdersEmailSettings",
+            payload: state
+        })
```
Comment (Tiggles) [inline frontend-web/webclient/app/UserSettings/ChangeEmailSettings.tsx:170]: Missing `;`

File: frontend-web/webclient/app/UserSettings/ChangeUserDetails.tsx [hunk 18]
```diff
@@ -18,7 +18,7 @@ const initialState: UserDetailsState = {
     placeHolderEmail: "Enter Email"
 };
 
-type Action<T, B> = {type: T; payload: B};
+export type Action<T, B> = {type: T; payload: B};
```
Comment (Tiggles) [inline frontend-web/webclient/app/UserSettings/ChangeUserDetails.tsx:21]: Use `PayloadAction` instead. It's globally defined.

File: frontend-web/webclient/app/UserSettings/ChangeUserDetails.tsx [hunk 19]
```diff
@@ -29,6 +29,7 @@ const reducer = (state: UserDetailsState, action: UpdatePlaceholderFirstNames):
 };
 
 export const ChangeUserDetails: React.FunctionComponent<{setLoading: (loading: boolean) => void}> = () => {
+    console.log("rerender")
```
Comment (Tiggles) [inline frontend-web/webclient/app/UserSettings/ChangeUserDetails.tsx:32]: Remove

File: frontend-web/webclient/app/UserSettings/api.tsx [hunk 20]
```diff
@@ -20,7 +20,7 @@ export function listUserSessions(
         context: ""
     };
 }
-
+/*
```
Comment (Tiggles) [inline frontend-web/webclient/app/UserSettings/api.tsx:23]: Remove

**Review by Tiggles — CHANGES_REQUESTED** [review]
Some cleanup needed.

File: frontend-web/webclient/app/UserSettings/ChangeEmailSettings.tsx [hunk 21]
```diff
@@ -11,8 +11,8 @@ import stat = file.stat;
 import EmailSettings = mail.EmailSettings;
 import retrieveEmailSettings = mail.retrieveEmailSettings;
 import toggleEmailSettings = mail.toggleEmailSettings;
-import {Action} from "UserSettings/ChangeUserDetails";
 import HexSpin from "LoadingIcon/LoadingIcon";
+import {Action} from "redux";
```
Comment (Tiggles) [inline frontend-web/webclient/app/UserSettings/ChangeEmailSettings.tsx:15]: I don't think this is needed?

File: frontend-web/webclient/app/UserSettings/ChangeUserDetails.tsx [hunk 22]
```diff
@@ -18,7 +18,7 @@ const initialState: UserDetailsState = {
     placeHolderEmail: "Enter Email"
 };
 
-export type Action<T, B> = {type: T; payload: B};
+type Action<T, B> = {type: T; payload: B};
```
Comment (Tiggles) [inline frontend-web/webclient/app/UserSettings/ChangeUserDetails.tsx:21]: This is a PayloadAction

**Review by Tiggles — APPROVED** [review]
Some changes could be made, but it's fine to merge now.
